src/rgw/rgw_op.cc

```diff
@@ -6359,9 +6364,103 @@ void RGWCompleteMultipart::pre_exec()
   rgw_bucket_object_pre_exec(s);
 }
 
+static inline int
+try_sum_part_cksums(const DoutPrefixProvider *dpp,
+		    CephContext *cct,
+		    rgw::sal::MultipartUpload* upload,
+		    RGWMultiCompleteUpload* parts,
+		    std::optional<int>& num_parts,
+		    std::optional<rgw::cksum::Cksum>& out_cksum,
+		    optional_yield y)
+{
+  /* 1. need checksum-algorithm header (if invalid, fail)
+     2. conditional on have-checksum,
+     3. need digest for supplied algo
+     4. iterate over parts, confirm each has same algo, if not, fail
+     5. for each part-checksum, accumlate bytes into new checksum
+     6. return armored and append "-<nparts>"
+     7. verify -- if invalid, fail */
+
+    /* rgw_sal.h says that list_parts is called for the side effect of loading
+     * the parts of an upload into "cache"--the api is strange and truncated
+     * flag suggests that it needs to be called multiple times to handle large
+     * uploads--but does not explain how that affects the hidden cache;  I'm
+     * assuming it turns over? */
+
+  int op_ret = 0;
+  bool truncated = false;
+  int marker = 0;
+
+  rgw::cksum::Type cksum_type{rgw::cksum::Type::none};
+  rgw::cksum::Cksum* prev_cksum{nullptr};
+
+  /* XXX as presently, it appears that if 'truncated' is passed as 'true' to
+   * list_parts, its value is never updated */
```

Comment:
Odd.  It looks like 'truncated' is always set and never read in RadosStore.

Reply:
I also missed resetting it to false in the truncated case, just fixed that